pkg/controllers/controllers.go

```diff
@@ -0,0 +1,176 @@
+	if len(systemNamespace) == 0 {
+		return errors.New("cannot start controllers on system namespace: system namespace not provided")
+	}
+
+	appCtx, err := newContext(cfg, systemNamespace)
+	if err != nil {
+		return err
+	}
+
+	chart.Register(ctx,
+		appCtx.K8s,
+		appCtx.Apply,
+		appCtx.HelmChart(),
+		appCtx.HelmChartConfig(),
+		appCtx.Batch.Job(),
+		appCtx.RBAC.ClusterRoleBinding(),
+		appCtx.Core.ServiceAccount(),
+		appCtx.Core.ConfigMap())
+
+	klog.Infof("Starting helm controller with %d threads.", opts.Threadiness)
+
+	if len(systemNamespace) == 0 {
+		klog.Info("Starting helm controller with no namespace.")
+	} else {
+		klog.Infof("Starting helm controller in namespace: %s.", systemNamespace)
+	}
+
+	leader.RunOrDie(ctx, systemNamespace, "helm-controller-lock", appCtx.K8s, func(ctx context.Context) {
+		if err := appCtx.start(ctx); err != nil {
+			klog.Fatal(err)
+		}
+		klog.Info("All controllers have been started")
+	})
+
+	return nil
+}
+
+func controllerFactory(rest *rest.Config) (controller.SharedControllerFactory, error) {
+	rateLimit := workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 60*time.Second)
+	workqueue.DefaultControllerRateLimiter()
+	clientFactory, err := client.NewSharedClientFactory(rest, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	cacheFactory := cache.NewSharedCachedFactory(clientFactory, nil)
+	return controller.NewSharedControllerFactory(cacheFactory, &controller.SharedControllerFactoryOptions{
+		DefaultRateLimiter: rateLimit,
+		DefaultWorkers:     50,
+	}), nil
+}
+
+func newContext(cfg clientcmd.ClientConfig, systemNamespace string) (*appContext, error) {
+	client, err := cfg.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	client.RateLimiter = ratelimit.None
+
+	apply, err := apply.NewForConfig(client)
+	if err != nil {
+		return nil, err
+	}
+	apply = apply.WithSetOwnerReference(false, false)
+
+	k8s, err := kubernetes.NewForConfig(client)
+	if err != nil {
+		return nil, err
+	}
+
+	scf, err := controllerFactory(client)
+	if err != nil {
+		return nil, err
+	}
+
+	core, err := core.NewFactoryFromConfigWithOptions(client, &core.FactoryOptions{
+		SharedControllerFactory: scf,
+	})
+	if err != nil {
+		return nil, err
+	}
+	corev := core.Core().V1()
+
+	clientConfig, err := cfg.ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientConfig.RateLimiter = ratelimit.None
+
+	helm, err := helm.NewFactoryFromConfigWithNamespace(clientConfig, systemNamespace)
+	if err != nil {
+		klog.Fatalf("Error building sample controllers: %s", err.Error())
+	}
+	helmv := helm.Helm().V1()
+
+	batch, err := batch.NewFactoryFromConfigWithNamespace(clientConfig, systemNamespace)
+	if err != nil {
+		klog.Fatalf("Error building sample controllers: %s", err.Error())
+	}
+	batchv := batch.Batch().V1()
+
+	rbac, err := rbac.NewFactoryFromConfigWithNamespace(clientConfig, systemNamespace)
+	if err != nil {
+		klog.Fatalf("Error building sample controllers: %s", err.Error())
+	}
```

Comment:
The errors here all say "sample controllers" which feels like copypasta from an example somewhere; should these instead indicate which specific controller factory they were trying to create?